### ISSUE TITLE
Incorrect type inferred when inner class definition closes over variable from outer class’s method

### ISSUE THREAD

**Issue by Hawk777** [PyCQA/pylint#8367]
### Bug description

This code is valid and works fine when executed, but fails Pylint checking:
```python
class Outer:
    def __init__(self):
        pass

    def say_hello(self):
        print("Hello World!")

    def make_inner(self):
        this = self
        class Inner:
            def get_outer(self):
                return this
        return Inner()


x = Outer().make_inner().get_outer()
x.say_hello()
```

### Command used

```shell
pylint hello.py --disable=missing-{module,class,function}-docstring --disable=too-few-public-methods
```

### Pylint output

```shell
************* Module hello
hello.py:17:0: E1101: Instance of 'Inner' has no 'say_hello' member (no-member)

------------------------------------------------------------------
Your code has been rated at 6.15/10 (previous run: 5.38/10, +0.77)
```

### Expected behavior

The `self` parameter to `make_inner` is of type `Outer`, therefore the closed-over `this` variable is of type `Outer`, therefore the return value of `get_outer` is of type `Outer`, therefore `x` is of type `Outer`.

### Pylint version

```shell
pylint 2.16.2
astroid 2.14.2
Python 3.10.10 (main, Feb 28 2023, 22:13:14) [GCC 12.2.1 20230121]
```

### OS / Environment

Gentoo Linux, in a venv
